src/atomics/relaxed.md

```diff
@@ -0,0 +1,423 @@
+            └────┘              ┃              └────┘
+```
+
+Note that **both sides must be atomic to avoid the data race**: if only the
+writing side used atomic operations, the reading side would still have multiple
+values to choose from (UB), and if only the reading side used atomic operations
+it could end up reading the garbage data “in-between” `0` and `1` (also UB).
+
+> **NOTE:** This description of why both sides are needed to be atomic
+> operations, while neat and intuitive, is not strictly correct: in reality the
+> answer is simply “because the spec says so”. However, it is isomorphic to the
+> real rules, so it can aid in understanding.
+
+## Read-modify-write operations
+
+Loads and stores are pretty neat in avoiding data races, but you can’t get very
+far with them. For example, suppose you wanted to implement a global shared
+counter that can be used to assign unique IDs to objects. Naïvely, you might try
+to write code like this:
+
+```rust
+# use std::sync::atomic::{self, AtomicU64};
+static COUNTER: AtomicU64 = AtomicU64::new(0);
+pub fn get_id() -> u64 {
+    let value = COUNTER.load(atomic::Ordering::Relaxed);
+    COUNTER.store(value + 1, atomic::Ordering::Relaxed);
+    value
+}
+```
+
+But then calling that function from multiple threads opens you up to an
+execution like below that results in two threads obtaining the same ID:
+
+```text
+Thread 1   COUNTER   Thread 2
+╭───────╮   ┌───┐   ╭───────╮
+│ load  ├───┤ 0 ├───┤  load │
+╰───╥───╯   └───┘   ╰────╥──╯
+╭───⇓───╮ ┌─┬───┐   ╭────⇓──╮
+│ store ├─┘ │ 1 │ ┌─┤ store │
+╰───────╯   └───┘ │ ╰───────╯
+            ┌───┬─┘
+            │ 1 │
+            └───┘
+```
+
+Technically, I believe it is _possible_ to implement this kind of thing with
+just loads and stores, if you try hard enough and use several atomics. But
+luckily, you don’t have to because there also exists another kind of operation,
+the read-modify-write, which is specifically suited to this purpose.
+
+A read-modify-write operation (shortened to RMW) is a special kind of atomic
+operation that reads, changes and writes back a value _in one step_. This means
+that there are guaranteed to exist no other values in the modification order in
+between the read and the write; it happens as a single operation. I would also
+like to point out that this is true of **all** atomic orderings, since a common
+misconception is that the `Relaxed` ordering somehow negates this guarantee.
+
+There are many different RMW operations to choose from, but the one most
+appropriate for this use case is `fetch_add`, which adds a number to the atomic,
+as well as returns the old value. So our code can be rewritten as this:
+
+```rust
+# use std::sync::atomic::{self, AtomicU64};
+static COUNTER: AtomicU64 = AtomicU64::new(0);
+pub fn get_id() -> u64 {
+    COUNTER.fetch_add(1, atomic::Ordering::Relaxed)
+}
+```
+
+And then, no matter how many threads there are, that race condition from earlier
```

Comment:
Was the distinction between a race condition and a data race ever specified? I may have gotten my words mixed up there but IIRC one is just unspecified but defined behavior, but the other is UB.

Reply:
That's right